inc/Engine/CriticalPath/Admin/Admin.php

```diff
@@ -0,0 +1,277 @@
+
+		$cpcss_pending = get_transient( 'rocket_cpcss_generation_pending' );
+
+		if ( ! empty( $cpcss_pending ) ) {
+			$cpcss_pending = $this->process_cpcss_pending_queue( (array) $cpcss_pending );
+		}
+
+		if ( false !== $cpcss_pending && empty( $cpcss_pending ) ) {
+			delete_transient( 'rocket_cpcss_generation_pending' );
+		}
+
+		if ( empty( $cpcss_pending ) ) {
+			$this->generation_complete();
+			wp_send_json_success( [ 'status' => 'cpcss_complete' ] );
+
+			return;
+		}
+
+		set_transient( 'rocket_cpcss_generation_pending', $cpcss_pending, HOUR_IN_SECONDS );
+		wp_send_json_success( [ 'status' => 'cpcss_running' ] );
+	}
+
+	/**
+	 * Pull one item off of the CPCSS Pending Queue and process it.
+	 *
+	 * @since 3.6
+	 *
+	 * @param array $cpcss_pending CPCSS Pending Queue.
+	 *
+	 * @return array remaining queue.
+	 */
+	private function process_cpcss_pending_queue( array $cpcss_pending ) {
+		$cpcss_item = reset( $cpcss_pending );
+		if ( empty( $cpcss_item ) ) {
+			return $cpcss_pending;
+		}
+
+		// Threshold 'check' > 10 = timed out.
+		$timeout          = ( $cpcss_item['check'] > 10 );
+		$cpcss_generation = $this->processor->process_generate(
+			$cpcss_item['url'],
+			$cpcss_item['path'],
+			$timeout,
+			! empty( $cpcss_item['mobile'] ) ? (bool) $cpcss_item['mobile'] : false
+		);
+
+		// Increment this item's threshold count.
+		$cpcss_pending[ $cpcss_item['path'] ]['check']++;
+
+		$this->cpcss_heartbeat_notices( $cpcss_generation, $cpcss_item );
+
+		// Remove the item from the queue when (a) the CPCSS API returns success or error or (b) timeouts.
+		if (
+			is_wp_error( $cpcss_generation )
+			||
+			'cpcss_generation_successful' === $cpcss_generation['code']
+			||
+			'cpcss_generation_failed' === $cpcss_generation['code']
+			||
+			$timeout
+		) {
+			unset( $cpcss_pending[ $cpcss_item['path'] ] );
+		}
+
+		return $cpcss_pending;
+	}
+
+	/**
+	 * CPCSS heartbeat update notices transients.
+	 *
+	 * @since 3.6
+	 *
+	 * @param array|WP_Error $cpcss_generation CPCSS regeneration reply.
+	 * @param array          $cpcss_item       Item processed.
+	 */
+	private function cpcss_heartbeat_notices( $cpcss_generation, $cpcss_item ) {
+		$mobile    = isset( $cpcss_item['mobile'] ) ? $cpcss_item['mobile'] : 0;
+		$transient = (array) get_transient( 'rocket_critical_css_generation_process_running' );
+
+		// Initializes the transient.
+		if ( ! isset( $transient['items'] ) ) {
+			$transient['items'] = [];
+		}
+
+		if ( is_wp_error( $cpcss_generation ) ) {
+			$item_path = ! (bool) $mobile ? $cpcss_item['path'] : str_replace( '-mobile.css', '.css', $cpcss_item['path'] );
+
+			$transient['items'][ $item_path ]['status'][ ! (bool) $mobile ? 'nonmobile' : 'mobile' ]['message'] = $cpcss_generation->get_error_message();
+			$transient['items'][ $item_path ]['status'][ ! (bool) $mobile ? 'nonmobile' : 'mobile' ]['success'] = false;
+			set_transient( 'rocket_critical_css_generation_process_running', $transient, HOUR_IN_SECONDS );
```

Comment:
@crystinutzaa This code and the code on lines 155 to 159 can be refactored into a private method. 

```php
private function set_process_running_transient( $mobile, $item_path, $cpcss_item, $cpcss_generation, $success = false ) {
	$item_path = ! (bool) $mobile ? $cpcss_item['path'] : str_replace( '-mobile.css', '.css', $cpcss_item['path'] );

	$transient['items'][ $item_path ]['status'][ ! (bool) $mobile ? 'nonmobile' : 'mobile' ]['message'] = $cpcss_generation->get_error_message();
	$transient['items'][ $item_path ]['status'][ ! (bool) $mobile ? 'nonmobile' : 'mobile' ]['success'] = $success;

	set_transient( 'rocket_critical_css_generation_process_running', $transient, HOUR_IN_SECONDS );
}